fix mrp calculation by use of low level code

scosist opened the pull request:
Fixes #3626 by generating a low level code to be used when calculating reorder rules.

Although the problem exists with 8.0 (see issue for current and expected behavior), the solution requires data model changes so the PR is on master with the intention that it will be accepted and then merged into 8.0.

scosist commented:
CLA provided by #6455

matt454357 commented:
Maybe @pedrobaeza and the @odoomrp team would be more interested in this PR?

pedrobaeza commented:
This area needs a lot of improvements, starting without a breaking on the atomicity of calculation methods, to provide some hooks for any change. This LLC can be a way of changing things afterwards, so it's very welcome, but in my opinion it's not enough. We can work together to improve this area for v9 if people from Odoo can give us a little guarantee that this work is going to land on it.

matt454357 commented:
Thanks for the reply Pedro.  We need to fix things up for v8, regardless of Odoo's decision on v9.  Can you offer advice on how to share our work?  Would @odoomrp be interested in a pull request, containing a module for v8?  We are happy to add comments, and better explain our solution.

pedrobaeza commented:
Yes, of course, please make the PR and I'll review it ASAP.

mart-e commented:
As discussed on #3626, I don't think it's the most efficient way to fix this issue. For people with many complex and sub-BoM, it may make sense to have some automation but for most use cases, a sequence field is probably enough.
Also BoM components is probably not the only criteria that should be considered for the order.

antonylesuisse commented:
The issue is similar to pricelist with hierchical categories https://github.com/odoo/odoo/pull/982, i dont want any complex mecanism. I would accept a patch that add a sequence (like we have in pricelists).

Then you could add a base_action_rule to compute the sequence based on boms if you wanted.

matt454357 commented:
@antonylesuisse, thank you for your recommendation, but why did you close this?  Do you want us to start a new pull request, with your recommended solution?

I am happy to rename the "llc" field to "sequence".  However, moving the computation to base_action_rule will be very inefficient.  That would require a recalculation of the sequence for every ```stock.warehouse.orderpoint``` record upon any change to ```mrp.bom.line```.  Ideally, the low-level-code should only be computed when starting the Run Schedulers operation.

Perhaps we could compromise by moving the LLC computation to a new module?

As a side note, every competing software includes LLC as a core and essential part of MRP calculation.  It is a fundamental concept in Material Requirements Planning.  I understand that some users will not have multi-level bills of materials (i.e. sub-assemblies), but in manufacturing, I believe most will.  **Even the demo data has sub-assemblies.**

antonylesuisse commented:
@jco-odoo and @fpodoo are working on the mrp refactoring for v10 i would like to get their opinion first.

scosist commented:
I implemented the changes described by @matt454357. We changed from using a recursive ORM method to creating a database view to calculate the LLC. The query Matt provided offers a significant improvement in performance over previous attempts using the ORM. This PR adds a sequence field indicated by @antonylesuisse and moves the LLC computation to an optional module.

antonylesuisse commented:
@jco-odoo sequence for the mrp branch, the module should go in apps.

jco-odoo commented on addons/mrp_llc/mrp.py:
```diff
@@ -0,0 +1,84 @@
+        llc_obj = self.env['mrp.bom.llc']
+        llc_ids = llc_obj.search([])
+        for llc_id in llc_ids:
+            for orderpoint in self.env['stock.warehouse.orderpoint'].search(
```
a simple browse would be less confusing

jco-odoo commented:
I am excited we can have such pull requests and it is a pity it escaped my eyes for such a long time.  

Normally, I would have said that the solution is to run the schedulers multiple times until there is no orderpoints to fulfill anymore.  That would also require merging production orders (which we will make possible, but it is cleaner when you can avoid it).  This solution however would allow to do all orderpoints in one pass.  

I see some small catches: 
- It does not take into account  the order e.g. when one would do resupply between warehouses.  This could however be a second sequence field: when you have multiple orderpoints of the same product it checks the procurement rules to see which ones need to be executed first.  
- I don't know why so much effort is put into phantom BoMs in the query as there are no orderpoints of them.  It is the same with MTO+manufacture products normally.  I need to study the query further.

jco-odoo commented:
Actually the sequence field is a little bit like a parent_left, parent_right and as you said, the user does not see it.

antonylesuisse commented:
@jco-odoo that was the inital pr, but i asked them explicitly to split the ordering in a separate module.

matt454357 commented:
@jco-odoo, Thanks,

> It does not take into account the order e.g. when one would do resupply between warehouses. This could however be a second sequence field: when you have multiple orderpoints of the same product it checks the procurement rules to see which ones need to be executed first.

That's fine with us.  Let's make an llc field and a sequence field for location.  All orderpoints for a given product will have the same llc, and they are all going to get executed.  We just need to make sure we sort by llc first.

Our current code: ```order="sequence, location_id"```
Proposed code: ```order="llc, sequence"```

The sequence field should be automatically populated, based on material flow between locations, but I'm not ready to write code for that.

>And I am not sure we really need a field. If we have just one query returning the order, that might do the trick as well. (just being careful not to forget any).

I'm not sure how that would work.  A query that returns all of stock.warehouse.orderpoint would be faster, but we need this query in multiple places, so wouldn't we want this in a model?  Prototype inheritance would be messy for fields added in the future (or computed).

matt454357 commented:
@jco-odoo,

> I don't know why so much effort is put into phantom BoMs in the query

Yes, I agree.  I can not think of a scenario where we need to worry about children of phantoms for LLC.  I will simplify the query.

matt454357 commented:
@jco-odoo, Here's the simplified query.  If you like it, we will update the PR.

```sql
with j as (
    -- if only template specified on mrp.bom, then expand list with all variants
    SELECT DISTINCT
        COALESCE(b.product_id,p.id) AS parent_id,
        l.product_id AS comp_id
    FROM mrp_bom_line AS l, mrp_bom AS b, product_product AS p
    WHERE b.product_tmpl_id=p.product_tmpl_id
    AND l.bom_id=b.id
)
SELECT *
FROM (
    -- build a path array
    with recursive stack(parent_id, comp_id, path) as (
        SELECT
            j.parent_id,
            j.comp_id,
            ARRAY[j.parent_id, j.comp_id]
        FROM j
        WHERE j.parent_id NOT IN (SELECT comp_id FROM j)
        UNION ALL
        SELECT
            j.parent_id,
            j.comp_id,
            path || j.comp_id
        FROM stack AS s, j
        WHERE j.parent_id=s.comp_id
    )
    -- use longest path for each orderpoint as llc
    SELECT
        op.id,
        COALESCE(MAX(ARRAY_LENGTH(path, 1))-1, 0) AS llc
    FROM stock_warehouse_orderpoint AS op
    LEFT JOIN stack AS s ON op.product_id=s.comp_id
    GROUP BY op.id
) AS res
```

matt454357 commented:
This has another small improvement:

```sql
with j as (
    -- if only template specified on mrp.bom, then expand list with all variants
    SELECT DISTINCT
        COALESCE(b.product_id,p.id) AS parent_id,
        l.product_id AS comp_id,
        b.type
    FROM mrp_bom_line AS l, mrp_bom AS b, product_product AS p
    WHERE b.product_tmpl_id=p.product_tmpl_id
    AND l.bom_id=b.id
)
SELECT *
FROM (
    -- build a path array
    with recursive stack(parent_id, comp_id, path) as (
        SELECT
            j.parent_id,
            j.comp_id,
            ARRAY[j.comp_id]
        FROM j
        WHERE j.parent_id NOT IN (SELECT comp_id FROM j)
        UNION ALL
        SELECT
            j.parent_id,
            j.comp_id,
            path || j.comp_id
        FROM stack AS s, j
        WHERE j.parent_id=s.comp_id
    )
    -- use longest path for each orderpoint as llc
    SELECT
        op.id,
        COALESCE(MAX(ARRAY_LENGTH(path, 1)), 0) AS llc
    FROM stock_warehouse_orderpoint AS op
    LEFT JOIN stack AS s ON op.product_id=s.comp_id
    GROUP BY op.id
) AS res
```

matt454357 commented:
@jco-odoo,

I was mistaken here:
> We just need to make sure we sort by llc first.
> 
> Our current code: order="sequence, location_id"
> Proposed code: order="llc, sequence"

Without considering resupply between warehouses, we would sort by location_id first, then llc:
```order="location_id, llc"```
This satisfies my need (I'm not using resupply).

I believe that resupply between warehouses is a separate issue, but I have some thoughts on that topic.

tivisse commented:
@antonylesuisse @jco-odoo 

It seems you've participated actively in this PR for the long time, even before the MRP refactoring. What is the status here ? Has it been integrated into the new MRP so we can close this PR ? It is still relevant ? Should we review it ? 

Thanks.

jco-odoo commented:
It is a very interesting feature.  But others took priority.

On Fri, Apr 28, 2017 at 10:41 AM, Tivisse <notifications@github.com> wrote:

> @antonylesuisse <https://github.com/antonylesuisse> @jco-odoo
> <https://github.com/jco-odoo>
>
> It seems you've participated actively in this PR for the long time, even
> before the MRP refactoring. What is the status here ? Has it been
> integrated into the new MRP so we can close this PR ? It is still relevant
> ? Should we review it ?
>
> Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/odoo/odoo/pull/4315#issuecomment-297942404>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADp7e0hgttmEHpbSNZdaOHvlfNR3bt-qks5r0aY3gaJpZM4DJ6Q9>
> .
>

tde-banana-odoo commented:
@tivisse As you took the lead on this, any news ?

mart-e commented:
@jco-odoo so can we give a final comment? If it is no longer a priority we can reject it but if you believe it will be useful, maybe we can merge it now

tivisse commented:
It's been a long time that this PR is opened, and neither @antonylesuisse nor @jco-odoo could take a decision about this new feature. 

I'm gonna close it, and set it in our internal feedback process. Our Product Owners team will qualify this and decide whether we do it or not.

Thanks again for the feedback.

Yannick.
